nixos/modules/security/gnupg.nix

```diff
@@ -0,0 +1,271 @@
+  # GPG_AGENT_INFO can no longer be used as it is ignored with gpg >= 2.1,
+  # hence three different hacks are done here to make gpg connect
+  # to ${gnupgHome}/S.gpg-agent depending on the concern:
+  # - For gpg-agent, --supervised mode is used to pass it a socket
+  #   in the persistent directory gnupgHome.
+  # - For the root user, on its login pam_systemd is mounting a fresh tmpfs on /run/user/0
+  #   so wrong perms are set on /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
+  #   when /run/user/0 is mounted, by overriding user-runtime-dir@.service
+  # - For secret decrypting services, /run/user/0/gnupg
+  #   is emptied and keept empty by privately mounting
+  #   an empty directory on it, using BindReadOnlyPaths=
+  systemd.sockets."gpg-agent" = lib.optionalAttrs cfg.enableGpgAgent {
+    description = "Socket for gpg-agent";
+    wantedBy = ["sockets.target"];
+    socketConfig.ListenStream = "${gnupgHome}/S.gpg-agent";
+    socketConfig.SocketMode = "0600";
+  };
+  environment.systemPackages =
+    # TODO: maybe this would be better to do that directly in pkgs.gnupg?
+    let gpgPresetPassphrase = pkgs.runCommand "gpg-preset-passphrase"
+      { preferLocalBuild = true;
+        allowSubstitutes = false;
+      } ''
+      mkdir -p $out/bin
+      ln -s -t $out/bin ${pkgs.gnupg}/libexec/gpg-preset-passphrase
+    ''; in
+    [ pkgs.gnupg gpgPresetPassphrase ];
+  systemd.packages = [
+    # Here, passing by systemd.packages is kind of a hack to be able to
+    # write this file which is neither writable using environment.etc
+    # (because environment.etc."systemd/system".source is set)
+    # nor using systemd.services (because systemd.services."user-runtime-dir@0"
+    # does not exist, and should not to keep using systemd's upstream template
+    # and systemd.services."user-runtime-dir@").
+    (pkgs.writeTextDir "etc/systemd/system/user-runtime-dir@0.service.d/override.conf" ''
+      [Unit]
+      [Service]
+      ExecStartPost=${pkgs.writeShellScript "redirect-gpg-agent-run-socket" ''
+        install -D -d -m 640 /run/user/0/gnupg/d.6qoenf9br6fajbkknuz1i6ts
+      ''}
+    '')
+  ];
+  systemd.services =
+    lib.optionalAttrs cfg.enableGpgAgent {
+      gpg-agent = {
+        description = "gpg-agent for decrypting GnuPG-protected secrets";
+        requires = ["gpg-agent.socket"];
+        serviceConfig = {
+          Type = "simple";
+          ExecStart = ''${pkgs.gnupg}/bin/gpg-agent \
+           --supervised \
+           --homedir '${gnupgHome}' \
+           --allow-loopback-pinentry \
+           --allow-preset-passphrase \
+           ${lib.escapeShellArgs cfg.gpgAgentFlags}
+          '';
+          Restart = "on-failure";
+          RestartSec = 5;
+          StateDirectory = ["gnupg" "gnupg/empty"];
+          StateDirectoryMode = "700";
+        };
+      };
+    } //
+    lib.mapAttrs' (target: secret:
+      lib.nameValuePair (lib.removeSuffix ".service" secret.service) {
```

Comment:
So does this service starts a new gnupg daemon? Because it bind mounts `/var/lib/gnupg/empty`?